### PR TITLE
fix: if response.setHeader is used, will cause header write after send error

### PR DESCRIPTION
**Pull request by xtx1130**
If use `reply.res.writeHead` before `reply.send(stream)`, it will cause `response.headersSent` to be true, and `res.setHeader` will throw a error.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
Can you please add a unit test for your change?

**Comment by xtx1130** [comment]
PTAL

File: test/internals/reply.test.js [hunk 1]
```diff
@@ -293,6 +293,38 @@ test('buffer with content type should not send application/octet-stream', t => {
...
 })
 
+test('stream using reply.res.writeHead should return customize headers', t => {
+  t.plan(4)
+
+  const fastify = require('../..')()
+  const fs = require('fs')
+  const path = require('path')
+
+  let streamPath = path.join(process.cwd(), './package.json')
```
Comment (mcollina) [inline test/internals/reply.test.js]: Can you use path.join(__dirname, ‘..’, ‘package.json’) instead? it’s more robust.

Comment (delvedor) [reply test/internals/reply.test.js]: Can you also use `var` instead of `let`?

Comment (delvedor) [reply test/internals/reply.test.js]: General speaking `let` is slower, but in this case for consistency across the test :)

Comment (xtx1130) [reply test/internals/reply.test.js]: @delvedor OK, I got it :)

File: lib/reply.js [hunk 2]
```diff
@@ -249,8 +249,10 @@ function sendStream (payload, res, reply) {
   // appropriately, e.g. a 404 for a missing file. So we cannot use
   // writeHead, and we need to resort to setHeader, which will trigger
   // a writeHead when there is data to send.
-  for (var key in reply._headers) {
-    res.setHeader(key, reply._headers[key])
+  if (!res.headersSent) {
```
Comment (delvedor) [inline lib/reply.js]: Should we add a warn log here? Maybe inside an else branch.
cc @mcollina

File: lib/reply.js [hunk 3]
```diff
@@ -249,8 +249,12 @@ function sendStream (payload, res, reply) {
...
+    for (var key in reply._headers) {
+      res.setHeader(key, reply._headers[key])
+    }
+  } else {
+    res.log.warn('response will send, but you shouldn\'t use res.writeHead in stream mode')
```
Comment (xtx1130) [inline lib/reply.js]: I have added a warn, but I'm not sure if the message is ok or not :)

File: test/internals/reply.test.js [hunk 4]
```diff
@@ -293,6 +293,41 @@ test('buffer with content type should not send application/octet-stream', t => {
   })
 })
 
+test('stream using reply.res.writeHead should return customize headers', t => {
```
Comment (allevo) [inline test/internals/reply.test.js]: The test confuses me a bit: no header are set, so there's no reason for call the warn.

The truth is that fastify adds an header internally here
https://github.com/fastify/fastify/blob/master/lib/reply.js#L57-L63

And that is why the warn is raised.

Maybe adding an explicit header using `res.header`  method, the test becomes more readable

Comment (xtx1130) [reply test/internals/reply.test.js]: I have separate my test case about stream mode, I think it will be more clearly.   

`res.setHeader` won't dispatch the else branch, only for `res.writeHead`.  If `res.writeHead` has called before `res.setHeader`and don't have a judgment about `res.headersSent`, it will throw an error.  And that is the reason why I add this judgment, It can stop the error happening.

File: lib/reply.js (hunk 3 above)
Comment (delvedor) [inline lib/reply.js]: I would rephrase this as:`res.log.warn('Headers already sent, the following headers will not be sent', reply._headers)`

Comment (xtx1130) [reply lib/reply.js]: I don't think so.   
Because when `response.headersSent === true`, it denotes the header has been send. Either `reply._headers` or `response.getHeader` can't get the value send by `reply.res.writeHead`

Comment (delvedor) [reply lib/reply.js]: So this case could happen **only** if you are using `writeHead`?

Comment (xtx1130) [reply lib/reply.js]: Yes :)
